psi4/src/read_options.cc

```diff
@@ -1408,6 +1408,8 @@ int read_options(const std::string &name, Options &options, bool suppress_printi
             orbitals before switching to the use of exact integrals in
             a |globals__scf_type| ``DIRECT`` calculation -*/
         options.add_bool("DF_SCF_GUESS", true);
+        /*- Forcibly disable or enable in- or out-of-core algorithms for MemDF/DiskDF. For debug purposes. !expert -*/
+        options.add_str("FORCE_MEM", "AUTO", "AUTO FORCE_INCORE NO_INCORE");
```

Comment:
I guess what troubles me about this PR is that it's adding a new option `FORCE_MEM` that pertains to one SCF_TYPE, but there's a `PK_NO_INCORE` already that does a partial job of the same but for a different single SCF_TYPE.

I wonder if there should be a single `INCORE_TYPE` option with values `AUTO`, `INCORE`, `OUTOFCORE` (or `MEM`/`DISK` or `FORCE_INCORE`/`NO_INCORE` or any combination thereof) and have it work for `PK` (will need more logic b/c mem-only not selectable at present), `MEM_DF`, `DISK_DF` (in future PR), and maybe `DF` (auto-selection of MEM_DF/DISK_DF)? What do folks think?

Reply:
Just as a note, this PR applies FORCE_MEM to both SCF_TYPE=MEM_DF and SCF_TYPE=DF (the latter when MemDFJK is automatically selected, of course). The implementation of FORCE_MEM in MemDFJK/DFHelper is independent of the SCF_TYPE used to utilize it.

I do see what you mean about overlapping keywords, though. This PR set is currently planned to be a two-part project, where FORCE_MEM will apply to both the MemDFJK and DiskDFJK algorithms by the end. Based on what you mentioned, it may be worth combining the PK_NO_INCORE keyword with the FORCE_MEM option so that all of MEM_DF, DISK_DF, DF, and PK use it. 

Such a plan should be doable across separate PRs, of which this PR is one. That said, this all does bring up the question of PR ordering. This PR introduces the FORCE_MEM option and implements it into MemDFJK, but maybe it is better to split those two and have the introduction of FORCE_MEM replace PK_NO_INCORE for the initial PR?

Reply:
what do you think of the below?
```
/*- For certain |globals__scf_type| algorithms that have internal sub-algorithms
    depending on available memory or other hardware constraints, allow the best
    sub-algorithm for the molecule and conditions (``AUTO`` ; usual mode) or
    forcibly select a sub-algorithm (usually only for debugging or profiling).
    Presently, ``SCF_TYPE=MEM_DF`` can have ``INCORE`` or ``OUT_OF_CORE`` selected.
    (This also applies for ``SCF_TYPE=DF`` when ``MEM_DF`` active.) !expert -*/
options.add_str("SCF_SUBTYPE", "AUTO", "AUTO INCORE OUT_OF_CORE");
```

Reply:
All right, after further discussion of the idea at the Psi4 dev meeting, I understand the scope of this suggestion better; and I will implement it!

Reply:
Done and done!